File: 문자열/사은품교환하기/1회차.py
'''
쿠폰 => 상품 교환
N 최소 5장
M 최소 7장 
총 12장 필요
단, N으로 부족한 개수 충족가능
최대 값은?
'''
# 풀이 법
'''
M//7가 클 때는
그냥 N//5값을 반환하는게 최적의 값이고
아닐 때는
두 값을 더한 다음 //12 를 해준다

왜냐면 50, 7일 때 45//12= 3, (5+7)//12=1
즉, 57//12 = 4

(6+6)/12 =1
'''
def f(N, M):
    # N, M = list(map(int, input().split(' ')))
    a = N//5
    b = M//7
    if b >= a:
      return a
    return (N+M)//12

    

# T = int(input())
# while T:
#     print(f())
#     T -= 1
print(f(12, 12), 2)
print(f(12, 13), 2)
print(f(6, 6), 1)
print(f(50, 7), 4)
print(f(500, 7), 42)
print(f(6, 6), 1)
print(f(12, 12), 2)
print(f(12, 13), 2)
print(f(15, 14), 2)
print(f(999, 0), 83)
print(f(5, 99999), 1)
print(f(6, 7), 1)
print(f(4, 20), 0)

File: leetcode/5.LongestPalindromicSubstring/1회차.js
//링크
/*
문제: https://leetcode.com/problems/longest-palindromic-substring/
가장 긴 팰린드롬(뒤집어도 같음 예: 토마토)을 주어진 문자열의 부분문자열에서 찾아라
참고:
*/

//풀이 법
/*
bruteforce 방식
문자열을 순회하며 모든 부분문자열을 가져온다.
만약 h에 없는 문자열이라면 h에 넣고
팰린드롬인지 판별한다.
만약 h에 있다면 스킵하고 다음 부분문자열을 찾는다.
*/

//키포인트
/*
성능 최적화
문자열길이가 최대 1000개 
bruteforce 방법은 n^3, 최대 n^2으로 줄여야함
*/

const isP = (S) => {
  const r = S.split('').reverse().join('')
  if (r === S) {
    return true
  }
  return false
}
function splice(S, i, j) {
  return S.split('').splice(i, j + 1 - i).join('')
}
var longestPalindrome = function (S) {
  const h = {}
  const l = S.length
  const results = []
  for (let i = 0; i < l; i++) {
    for (let j = i; j < l; j++) {
      const s = splice(S, i, j)
      if (h[s]) {
        continue;
      }
      h[s] = s
      if (isP(s)) {
        results.push(s)
      }
    }
  }

  results.sort((a, b) => b.length - a.length)
  ans = results[0]
  return ans
};


// longestPalindrome('abba')
// longestPalindrome("abcdbbfcba")
longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth")